backend/src/main/java/ai/verta/modeldb/datasetVersion/DatasetVersionDAORdbImpl.java

```diff
@@ -280,12 +300,101 @@ public String getUrlForDatasetVersion(String datasetVersionId, String method)
+    for (DatasetVersion datasetVersion : datasetVersionList) {
+      datasetIdDatasetVersionIdMap.put(datasetVersion.getId(), datasetVersion.getDatasetId());
+    }
+    Set<String> datasetIdSet = new HashSet<>(datasetIdDatasetVersionIdMap.values());
+
+    List<String> accessibleDatasetVersionIds = new ArrayList<>();
+    List<String> allowedDatasetIds;
+    // Validate if current user has access to the entity or not
+    if (datasetIdSet.size() == 1) {
+      roleService.isSelfAllowed(
+          ModelDBServiceResourceTypes.DATASET,
+          modelDBServiceActions,
+          new ArrayList<>(datasetIdSet).get(0));
+      accessibleDatasetVersionIds.addAll(requestedDatasetVersionIds);
+    } else {
+      allowedDatasetIds =
+          roleService.getSelfAllowedResources(
+              ModelDBServiceResourceTypes.DATASET, modelDBServiceActions);
+      // Validate if current user has access to the entity or not
+      allowedDatasetIds.retainAll(datasetIdSet);
+      for (Map.Entry<String, String> entry : datasetIdDatasetVersionIdMap.entrySet()) {
+        if (allowedDatasetIds.contains(entry.getValue())) {
+          accessibleDatasetVersionIds.add(entry.getKey());
+        }
+      }
+    }
+    return accessibleDatasetVersionIds;
+  }
+
   @Override
   public DatasetVersionDTO findDatasetVersions(
-      FindDatasetVersions queryParameters, UserInfo userInfo)
+      DatasetDAO datasetDAO, FindDatasetVersions queryParameters, UserInfo currentLoginUserInfo)
       throws InvalidProtocolBufferException {
     try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
 
+      List<String> accessibleDatasetVersionIds = new ArrayList<>();
+      if (!queryParameters.getDatasetVersionIdsList().isEmpty()) {
+        accessibleDatasetVersionIds.addAll(
+            getAccessibleDatasetVersionIDs(
+                queryParameters.getDatasetVersionIdsList(),
+                ModelDBActionEnum.ModelDBServiceActions.READ));
+        if (accessibleDatasetVersionIds.isEmpty()) {
+          String errorMessage =
+              "Access is denied. User is unauthorized for given DatasetVersion IDs : "
+                  + accessibleDatasetVersionIds;
+          ModelDBUtils.logAndThrowError(
+              errorMessage,
+              Code.PERMISSION_DENIED_VALUE,
+              Any.pack(FindDatasetVersions.getDefaultInstance()));
+        }
+      }
+
+      List<KeyValueQuery> predicates = new ArrayList<>(queryParameters.getPredicatesList());
+      for (KeyValueQuery predicate : predicates) {
+        if (predicate.getKey().equals(ModelDBConstants.ID)) {
+          List<String> accessibleDatasetVersionId =
+              getAccessibleDatasetVersionIDs(
+                  Collections.singletonList(predicate.getValue().getStringValue()),
+                  ModelDBActionEnum.ModelDBServiceActions.READ);
+          accessibleDatasetVersionIds.addAll(accessibleDatasetVersionId);
```

Comment:
Looks like this can be called twice, right? Here and above.

Reply:
Yes, @conradoverta, it is called twice but function `getAccessibleDatasetVersionIDs` accept first argument List of requested ids and based on that it will return only accessible ids from that requested ids so it is twice with different ids to check it is accessible or not.